### ISSUE TITLE
[fb] Numbers of grace notes stack above each other

### ISSUE THREAD

**Issue by WolfgangDrescher** [craigsapp/humlib#56]
Numbers look weird for grace notes. I think they should not at all be calculated.

http://verovio.humdrum.org/?file=beethoven/sonatas/sonata05-2.krn

`fb -c -a -3 --frequency 8`

<img width="697" alt="Bildschirm­foto 2023-01-16 um 10 30 10" src="https://user-images.githubusercontent.com/865594/212644359-b0c2665e-d7cc-4c45-a5ac-4040eb934e3a.png">

<img width="211" alt="Bildschirm­foto 2023-01-16 um 10 31 54" src="https://user-images.githubusercontent.com/865594/212644733-88028b1b-8971-4fcc-b368-c2123464c1a7.png">

**Comment by WolfgangDrescher** [comment]
Could a possible fix be to ignore notes with a `q` signifier?

```tsv
=17	=17	=17
.	(<ff8q	.
4An 4c 4e- 4f	4fff)	f
.	(eee-32qLLL	.
.	ccc32q	.
.	aan32q	.
.	ff32q	.
.	ee-32q	.
.	cc32q	.
.	an32qJJJ)	.
8r	8f	.
16.r	16.r	.
32AAn 32F	32c 32e-	p
```

**Comment by WolfgangDrescher** [comment]
Note that `<harm>` elements are created for each cue sized note with `fb -c -a -3 -n --frequency 4`. This is probably why there is a huge gap between the fb numbers and the score.

<img width="415" alt="Bildschirm­foto 2023-01-16 um 10 48 30" src="https://user-images.githubusercontent.com/865594/212648233-dea2f785-19ea-4124-bc03-c2f099120358.png">

<img width="176" alt="Bildschirm­foto 2023-01-16 um 10 49 46" src="https://user-images.githubusercontent.com/865594/212648512-bbb6edcf-2cfa-4cf0-af12-ce351c0b7165.png">

**Comment by craigsapp** [comment]
> Could a possible fix be to ignore notes with a q signifier?

It would be reasonable to ignore grace notes for fb analyses.   If you want to exclude them, there are two possibilities.

(1) You can ask the token for its duration (`token->getDuration()`) if the value is 0 then that means the note is a grace note.  Also the condition (token->find("q") != string::npos) also means a grace note (and it will have a duration of 0).

(2) If any note on a line has a `q`, then it can be inferred that all notes on the line also have `q` and a duration of 0.  If they don't, then I do not allow that (and the humlib parser will either complain or get very angry and crash when doing rhythm analysis.  You can ask the HumdrumLine object what its duration is with `HumdrumLine::getDuration()` if it reports 0, then that data line only contains grace note.  Example of asking the HumdrumLine from a token: `token->getOwner()->getDuration()`.   `HumdrumToken::getOwner()` returns a pointer to the HumdrumLine on which it is attached (or NULL if it is not in a HumdrumFile or HumdrumLine.

**Comment by craigsapp** [comment]
> Note that `<harm>` elements are created for each cue sized note with `fb -c -a -3 -n --frequency 4`. This is probably why there is a huge gap between the fb numbers and the score.

Note that grace  in a group all have the same timestamp.  So in order to display `<harm>` correctly on grace notes, the Humdrum-to-MEI converter (iohumdrum.cpp in the verovio repository) will probablyl need to be fix to not use `@tstamp` but rather `@startid`.

This fix would allow `<harm>` to be attached properly to grace note.  I had to do a similar enhancement for dynamics and hairpins in iohumdrum.cpp to all them to be attached to grace notes.  In other words, floating elements in MEI data require attaching things to grace notes with `@startd` rather than `@tstamp`.

**Comment by WolfgangDrescher** [comment]
(2) was a fast fix, thank you. I would have to test more why (1) did not work easily. Probably because NoteGrid  token resolving (and duration) seems to be a bit different from direct access to the tokens.

Implemented in bd09768. Since they are hidden now I won't need the `<harm>` position improvements in iohumdrum.cpp for now.

<img width="762" alt="Bildschirm­foto 2023-01-17 um 00 40 56" src="https://user-images.githubusercontent.com/865594/212780758-679e0edb-6146-47c0-902a-0204ee8bf9cb.png">
